packages/flutter/lib/src/rendering/debug_overflow_indicator.dart

```diff
@@ -0,0 +1,301 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:ui' as ui;
+
+import 'package:flutter/painting.dart';
+
+import 'object.dart';
+import 'stack.dart';
+
+// Data used by the DebugOverflowIndicator to manage the regions and labels for
+// the indicators.
+class _OverflowRegionData {
+  const _OverflowRegionData({
+    this.rect,
+    this.label = '',
+    this.labelOffset = Offset.zero,
+    this.rotation = 0.0,
+  });
+
+  final Rect rect;
+  final String label;
+  final Offset labelOffset;
+  final double rotation;
+}
+
+/// An mixin indicator that is drawn when a [RenderObject] overflows its
+/// container.
+///
+/// This is used by some RenderObjects that are containers to show where, and by
+/// how much, their children overflow their containers.  These indicators are
+/// typically only shown in a debug build (where the call to
+/// [paintOverflowIndicator] is surrounded by an assert).
+///
+/// This class will also print a debug message to the console when the container
+/// overflows, but only on the first occurrence, and after each time that
+/// [reassemble] is called.
+///
+/// ## Sample code
+///
+/// ```dart
+/// class MyRenderObject extends RenderAligningShiftedBox with DebugOverflowIndicatorMixin {
+///   MyRenderObject({
+///     RenderBox child,
+///     TextDirection textDirection,
+///     AlignmentGeometry alignment,
+///   }) : super.mixin() {
+///     this.textDirection = textDirection;
+///     this.alignment = alignment;
+///     this.child = child;
+///   }
+///   
+///   Rect _containerRect;
+///   Rect _childRect;
+///   bool _isOverflowing;
+///
+///   @override
+///   void performLayout() {
+///     // ...
+///     final BoxParentData childParentData = child.parentData;
+///     _containerRect = Offset.zero & size;
+///     _childRect = childParentData.offset & child.size;
+///   }
+///
+///   @override
+///   void paint(PaintingContext context, Offset offset) {
+///     if (!_isOverflowing) {  // (However you determine this for your container)
+///       // Do "normal" painting here and return, since there is no overflow.
+///       // ...
+///       return;
+///     }
+///     // ...
```

Comment:
maybe add a comment here saying to do the painting in case of overflow? it's not immediately clear from this sample why you test `_isOverflowing`; after all, one might expect the rendering to be the same codepath in both cases (it only isn't in practice because Flex clips overflow even in release builds).